File: compat/suites/node-js-sdk/src/groups/waf.ts
/**
 * groups/waf.ts — WAF v2 compatibility test groups for the Node.js suite.
 *
 * Status: NOT implemented in Overcast. All tests expected to fail with 501.
 * These tests define the coverage target for future WAF implementation.
 *
 * Groups:
 *   waf-webacls — Web ACL lifecycle
 */

import {
  CreateWebACLCommand,
  DeleteWebACLCommand,
  GetWebACLCommand,
  ListWebACLsCommand,
  Scope,
} from "@aws-sdk/client-wafv2";
import { makeClients } from "../lib/clients.js";
import type { TestGroup } from "../lib/harness.js";
import * as assert from "node:assert/strict";

export function makeWAFGroups(suite: string): TestGroup[] {
  return [
    // ── waf-webacls ────────────────────────────────────────────────────────
    {
      suite,
      service: "waf",
      name: "waf-webacls",
      tests: [
        {
          name: "CreateWebACL",
          fn: async (ctx) => {
            const { wafv2 } = makeClients(ctx);
            const name = `compat-${ctx.runId}`;
            const resp = await wafv2.send(
              new CreateWebACLCommand({
                Name: name,
                Scope: Scope.REGIONAL,
                DefaultAction: { Allow: {} },
                VisibilityConfig: {
                  SampledRequestsEnabled: false,
                  CloudWatchMetricsEnabled: false,
                  MetricName: `compat-${ctx.runId}`,
                },
                Rules: [],
              }),
            );
            assert.ok(resp.Summary?.Id, "CreateWebACL: missing Id");
            (ctx as Record<string, unknown>)["_aclId"] = resp.Summary.Id;
            (ctx as Record<string, unknown>)["_aclName"] = name;
            (ctx as Record<string, unknown>)["_aclLockToken"] =
              resp.Summary.LockToken;
          },
        },
        {
          name: "GetWebACL",
          fn: async (ctx) => {
            const { wafv2 } = makeClients(ctx);
            const id = (ctx as Record<string, unknown>)["_aclId"] as string;
            const name = (ctx as Record<string, unknown>)["_aclName"] as string;
            assert.ok(id || !name, "GetWebACL: no ACL from CreateWebACL");
            const resp = await wafv2.send(
              new GetWebACLCommand({
                Id: id,
                Name: name,
                Scope: Scope.REGIONAL,
              }),
            );
            assert.ok(resp.WebACL?.Id, "GetWebACL: missing Id");
          },
        },
        {
          name: "ListWebACLs",
          fn: async (ctx) => {
            const { wafv2 } = makeClients(ctx);
            await wafv2.send(new ListWebACLsCommand({ Scope: Scope.REGIONAL }));
          },
        },
        {
          name: "DeleteWebACL",
          fn: async (ctx) => {
            const { wafv2 } = makeClients(ctx);
            const id = (ctx as Record<string, unknown>)["_aclId"] as string;
            const name = (ctx as Record<string, unknown>)["_aclName"] as string;
            const lockToken = (ctx as Record<string, unknown>)[
              "_aclLockToken"
            ] as string;
            if (!id || !name || !lockToken) return;
            await wafv2.send(
              new DeleteWebACLCommand({
                Id: id,
                Name: name,
                Scope: Scope.REGIONAL,
                LockToken: lockToken,
              }),
            );
          },
        },
      ],
      teardown: async (ctx) => {
        const { wafv2 } = makeClients(ctx);
        const id = (ctx as Record<string, unknown>)["_aclId"] as string;
        const name = (ctx as Record<string, unknown>)["_aclName"] as string;
        if (!id || !name) return;
        try {
          // Refresh LockToken — it changes after each mutating call
          const fresh = await wafv2.send(
            new GetWebACLCommand({ Id: id, Name: name, Scope: Scope.REGIONAL }),
          );
          const lockToken = fresh.LockToken;
          if (!lockToken) return;
          await wafv2.send(
            new DeleteWebACLCommand({
              Id: id,
              Name: name,
              Scope: Scope.REGIONAL,
              LockToken: lockToken,
            }),
          );
        } catch {}
      },
    },
  ];
}

File: compat/suites/python-sdk/groups/waf.py
"""
groups/waf.py — WAF v2 compatibility test implementations for the Python suite.
"""

from __future__ import annotations
from lib.harness import TestContext
from lib.clients import make_clients


def _wafv2(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region)._get("wafv2")


# ── waf-webacls ───────────────────────────────────────────────────────────────

def CreateWebACL(ctx: TestContext) -> None:
    waf = _wafv2(ctx)
    name = f"compat-{ctx.run_id}"
    resp = waf.create_web_acl(
        Name=name,
        Scope="REGIONAL",
        DefaultAction={"Allow": {}},
        VisibilityConfig={
            "SampledRequestsEnabled": False,
            "CloudWatchMetricsEnabled": False,
            "MetricName": f"compat-{ctx.run_id}",
        },
        Rules=[],
    )
    summary = resp.get("Summary", {})
    if not summary.get("Id"):
        raise AssertionError("CreateWebACL: missing Id")
    ctx["waf_acl_id"] = summary["Id"]
    ctx["waf_acl_name"] = name
    ctx["waf_acl_lock_token"] = summary.get("LockToken", "")


def GetWebACL(ctx: TestContext) -> None:
    waf = _wafv2(ctx)
    acl_id = ctx.get("waf_acl_id")
    name = ctx.get("waf_acl_name")
    if not acl_id or not name:
        raise AssertionError("GetWebACL: no ACL from CreateWebACL")
    resp = waf.get_web_acl(Id=acl_id, Name=name, Scope="REGIONAL")
    if not resp.get("WebACL", {}).get("Id"):
        raise AssertionError("GetWebACL: missing Id")


def ListWebACLs(ctx: TestContext) -> None:
    _wafv2(ctx).list_web_acls(Scope="REGIONAL")


def DeleteWebACL(ctx: TestContext) -> None:
    waf = _wafv2(ctx)
    acl_id = ctx.get("waf_acl_id")
    name = ctx.get("waf_acl_name")
    lock_token = ctx.get("waf_acl_lock_token")
    if not acl_id or not name or not lock_token:
        return
    waf.delete_web_acl(
        Id=acl_id,
        Name=name,
        Scope="REGIONAL",
        LockToken=lock_token,
    )


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "CreateWebACL": CreateWebACL,
    "GetWebACL": GetWebACL,
    "ListWebACLs": ListWebACLs,
    "DeleteWebACL": DeleteWebACL,
}

SETUP = {}
TEARDOWN = {
    "waf-webacls": lambda ctx: _teardown_web_acl(ctx),
}


def _teardown_web_acl(ctx: TestContext) -> None:
    acl_id = ctx.get("waf_acl_id")
    name = ctx.get("waf_acl_name")
    lock_token = ctx.get("waf_acl_lock_token")
    if acl_id and name and lock_token:
        try:
            _wafv2(ctx).delete_web_acl(
                Id=acl_id,
                Name=name,
                Scope="REGIONAL",
                LockToken=lock_token,
            )
        except Exception:
            pass
